File: Concurrent_Toy_Language_Interpreter/View/TextMenu.java
package View;

import java.util.HashMap;
import java.util.Map;
import java.util.Scanner;

@SuppressWarnings("InfiniteLoopStatement")
class TextMenu {
    private Map<String, Command> commands;

    TextMenu() {
        this.commands = new HashMap<String, Command>();
    }

    void addCommand(Command cmd) {
        this.commands.put(cmd.getKey(), cmd);
    }

    private void printMenu() {
        for(Command cmd : this.commands.values()) {
            String line = String.format("%4s: %s", cmd.getKey(), cmd.getDescription());
            System.out.println(line);
        }
    }

    void show() {
        Scanner scanner = new Scanner(System.in);
        while(true) {
            this.printMenu();
            System.out.println("Input the option: ");
            String key = scanner.nextLine();
            Command com = commands.get(key);
            if(com == null) {
                System.out.println("Invalid option.");
                continue;
            }
            com.execute();
        }
    }
}

File: Toy_Language_Interpreter_C/View/TextMenu.cs
using System;
using System.Collections.Generic;
using System.Linq;
using System.Text;
using System.Threading.Tasks;

namespace a7
{
    class TextMenu
    {
        private Dictionary<String, Command> commands;

        public TextMenu()
        {
            this.commands = new Dictionary<string, Command>();
        }

        public void AddCommand(Command cmd)
        {
            this.commands.Add(cmd.Key, cmd);
        }

        private void PrintMenu()
        {
            foreach(Command cmd in this.commands.Values)
            {
                String line = String.Format("{0}: {1}", cmd.Key, cmd.Description);
                Console.WriteLine(line);
            }
            Console.WriteLine("\n");
        }

        public void Show()
        {
            while(true)
            {
                this.PrintMenu();
                String msg;
                Console.Write("Input the option: ");
                msg = Console.ReadLine();
                this.commands.TryGetValue(msg, out Command cmd);
                if(cmd == default(Command))
                {
                    Console.WriteLine("Invalid option.");
                    continue;
                }
                cmd.Execute();
            }

        }
    }
}
